### PR TITLE
Elastic training: Restart actors

### PR DESCRIPTION
**Pull request by krfricke**
Todo:
- docs (ENV variables)
- docs (behavior)

### CODE REVIEW

**Comment by krfricke** [comment]
Failing test is a timeout in `test_colocation.py::TestColocation::test_no_tune_spread` for latest ray master - this might actually be a regression from some recent PR. Could you take a look sometime @amogkam?
The relevant tests in this PR seem to be passing fine.

**Comment by amogkam** [comment]
OK will take a look

File: xgboost_ray/main.py
Comment (richardliaw) [inline xgboost_ray/main.py:517]: Can we move this + helper functions to a util file? Would it make sense to do that?

**Review by richardliaw — COMMENTED** [review]
Looks good, left a few minor nits (assuming it works!)

@amogkam can you please take a look too? Thanks!

**Review by amogkam — COMMENTED** [review]
This looks good to me overall! Just left some comments/questions.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:104]: I tried it and it didn't seem to make a difference. What would be the reasoning in making it a daemon?

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:1008]: This happens in the `_PrepareActorTask`:

```
        task = _PrepareActorTask(
            actor,
            queue=training_state.queue,
            stop_event=training_state.stop_event,
            load_data=load_data)
```

This is called in `_train()` and in `_maybe_schedule_new_actors` after creation of the actors.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:98]: Added a comment. It's basically the same reasoning as before, but there was an error - we were just putting the wait-for-join instruction into a process, not the tracker itself. By subclassing the RabitTracker we can actually kill the tracker.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:362]: The stop event gets overwritten every time we restart actors. When testing this PR I noticed that some training threads would continue to continue to run even after we stopped the `RayXGBoostActor.train()` function - there is no great way to kill threads, and putting this into a subprocess leads to a myriad of other problems (e.g. communication with the main process). Thus we need to make sure that training actually stops independently of the `RayXGBoostActor.train()` function, which is why I re-introduced the callback.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:369]: See comment above

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:553]: Yep, thanks!

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:843]: The check should prevent that we increase `restart_training_at`:
```
        if training_state.restart_training_at is None:
```

All actors that were promoted to proper actors will be restarted automatically as they are now in `training_state.actors`. As long as we keep them there we should be fine.

Actually, during clean up we should make sure to also shutdown pending actors. I'll add this.

File: xgboost_ray/main.py
Comment (krfricke) [reply xgboost_ray/main.py:517]: Yep, I thought about this, too. Moved it to `elastic.py` (and some things to `util.py`) - we'll need local imports now though to avoid circular imports

**Comment by krfricke** [comment]
The test on the cluster was successful:

```
(base) ray@ip-172-31-24-108:/release_tests$ python benchmark_cpu_gpu.py 4 1000 4                
2020-12-30 06:30:42,694 INFO worker.py:657 -- Connecting to existing Ray cluster at address: 172.31.24.108:6379
2020-12-30 06:30:42,910 WARNING worker.py:1034 -- The actor or task with ID ffffffffffffffff0ab974dd03000000 cannot be scheduled right now. It requires {actor_cpus: 4.000000}, {CPU: 4.000000} for placement, however the cluster currently cannot provide the requested resources
. The required resources may be added as autoscaling takes place or placement groups are scheduled. Otherwise, consider reducing the resource requirements of the task.
2020-12-30 06:30:42,910 WARNING worker.py:1034 -- The actor or task with ID ffffffffffffffffcee45dd103000000 cannot be scheduled right now. It requires {actor_cpus: 4.000000}, {CPU: 4.000000} for placement, however the cluster currently cannot provide the requested resources
. The required resources may be added as autoscaling takes place or placement groups are scheduled. Otherwise, consider reducing the resource requirements of the task.
2020-12-30 06:30:42,913 INFO main.py:721 -- [RayXGBoost] Created 4 new actors (4 total actors). Waiting until actors are ready for training.
2020-12-30 06:30:42,920 WARNING worker.py:1034 -- The actor or task with ID ffffffffffffffffbd328aee03000000 cannot be scheduled right now. It requires {actor_cpus: 4.000000}, {CPU: 4.000000} for placement, however the cluster currently cannot provide the requested resources
. The required resources may be added as autoscaling takes place or placement groups are scheduled. Otherwise, consider reducing the resource requirements of the task.
2020-12-30 06:30:42,921 WARNING worker.py:1034 -- The actor or task with ID ffffffffffffffff0703d58c03000000 cannot be scheduled right now. It requires {CPU: 4.000000}, {actor_cpus: 4.000000} for placement, however the cluster currently cannot provide the requested resources
. The required resources may be added as autoscaling takes place or placement groups are scheduled. Otherwise, consider reducing the resource requirements of the task.
2020-12-30 06:30:54,974 INFO main.py:758 -- [RayXGBoost] Starting XGBoost training.      
(pid=9436, ip=172.31.17.60) [06:30:54] task [xgboost.ray]:140427761209424 got new rank 0 
(pid=9129, ip=172.31.18.124) [06:30:54] task [xgboost.ray]:140607102729744 got new rank 1
(pid=9244, ip=172.31.22.246) [06:30:54] task [xgboost.ray]:139690855271952 got new rank 2
(pid=9113, ip=172.31.31.127) [06:30:54] task [xgboost.ray]:140379158221008 got new rank 3        
2020-12-30 06:31:07,268 INFO elastic.py:172 -- Actor status: 3 alive, 1 dead (4 total)           
2020-12-30 06:31:12,274 WARNING main.py:1050 -- A Ray actor died during training. Trying to continue training on the remaining actors. This will use 3 existing actors and start 0 new actors. Sleeping for 10 seconds for cleanup.
2020-12-30 06:31:22,292 INFO main.py:721 -- [RayXGBoost] Created 0 new actors (3 total actors). Waiting until actors are ready for training.
2020-12-30 06:31:22,398 INFO main.py:758 -- [RayXGBoost] Starting XGBoost training.               
(pid=9244, ip=172.31.22.246) [06:31:22] task [xgboost.ray]:139690855271952 got new rank 2         
(pid=9436, ip=172.31.17.60) [06:31:22] task [xgboost.ray]:140427761209424 got new rank 0          
(pid=9129, ip=172.31.18.124) [06:31:22] task [xgboost.ray]:140607102729744 got new rank 1         
2020-12-30 06:31:22,642 INFO elastic.py:91 -- Re-scheduled 1 actors for training. Once data loading finished, they will be integrated into training again.
2020-12-30 06:31:29,291 WARNING worker.py:1034 -- The actor or task with ID fffffffffffffffff9c9d3f303000000 cannot be scheduled right now. It requires {CPU: 4.000000}, {actor_cpus: 4.000000} for placement, but this node only has remaining {node:172.31.18.124: 1.000000}, {me
mory: 10.498047 GiB}, {object_store_memory: 3.076172 GiB}. In total there are 0 pending tasks and 1 pending actors on this node. This is likely due to all cluster resources being claimed by actors. To resolve the issue, consider creating fewer actors or increase the resource
s available to this Ray cluster. You can ignore this message if this Ray cluster is expected to auto-scale.                              
2020-12-30 06:31:52,416 INFO main.py:825 -- Training in progress (30 seconds since last restart). 
2020-12-30 06:32:22,416 INFO main.py:825 -- Training in progress (60 seconds since last restart).
2020-12-30 06:32:52,417 INFO main.py:825 -- Training in progress (90 seconds since last restart). 
2020-12-30 06:33:22,417 INFO main.py:825 -- Training in progress (120 seconds since last restart).
2020-12-30 06:34:22,418 INFO main.py:825 -- Training in progress (180 seconds since last restart).
2020-12-30 06:34:52,418 INFO main.py:825 -- Training in progress (210 seconds since last restart).
(pid=9113, ip=172.31.31.127) [06:30:54] task [xgboost.ray]:140379158221008 got new rank 3
(pid=9113, ip=172.31.31.127) 2020-12-30 06:31:07,197    ERROR worker.py:383 -- SystemExit was raised from the worker
(pid=9113, ip=172.31.31.127) Traceback (most recent call last):
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 550, in ray._raylet.task_execution_handler
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 417, in ray._raylet.execute_task
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 456, in ray._raylet.execute_task
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 459, in ray._raylet.execute_task
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 463, in ray._raylet.execute_task
(pid=9113, ip=172.31.31.127)   File "python/ray/_raylet.pyx", line 415, in ray._raylet.execute_task.function_executor
(pid=9113, ip=172.31.31.127)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/function_manager.py", line 556, in actor_method_executor
(pid=9113, ip=172.31.31.127)     return method(__ray_actor, *args, **kwargs)
(pid=9113, ip=172.31.31.127)   File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 462, in train
(pid=9113, ip=172.31.31.127)     raise RayXGBoostTrainingStopped("Training was interrupted.")
(pid=9113, ip=172.31.31.127)   File "/home/ray/anaconda3/lib/python3.7/site-packages/ray/worker.py", line 380, in sigterm_handler
(pid=9113, ip=172.31.31.127)     sys.exit(1)
(pid=9113, ip=172.31.31.127) SystemExit: 1
2020-12-30 06:35:11,945 INFO elastic.py:172 -- Actor status: 4 alive, 0 dead (4 total)
2020-12-30 06:35:12,213 INFO main.py:1043 -- A new actor became available. Re-starting training from latest checkpoint with new actor. This will use 4 existing actors and start 0 new actors. Sleeping for 10 seconds for cleanup.
2020-12-30 06:35:16,922 ERROR worker.py:980 -- Possible unhandled error from worker: ray::RayXGBoostActor.train() (pid=9129, ip=172.31.18.124)
  File "python/ray/_raylet.pyx", line 463, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 415, in ray._raylet.execute_task.function_executor
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 461, in train
    if self._stop_event.is_set():
xgboost_ray.main.RayXGBoostTrainingStopped: Training was interrupted.
2020-12-30 06:35:16,923 ERROR worker.py:980 -- Possible unhandled error from worker: ray::RayXGBoostActor.train() (pid=9436, ip=172.31.17.60)
  File "python/ray/_raylet.pyx", line 463, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 415, in ray._raylet.execute_task.function_executor
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 461, in train
    if self._stop_event.is_set():
xgboost_ray.main.RayXGBoostTrainingStopped: Training was interrupted.
2020-12-30 06:35:17,923 ERROR worker.py:980 -- Possible unhandled error from worker: ray::RayXGBoostActor.train() (pid=9244, ip=172.31.22.246)
  File "python/ray/_raylet.pyx", line 463, in ray._raylet.execute_task
  File "python/ray/_raylet.pyx", line 415, in ray._raylet.execute_task.function_executor
  File "/home/ray/anaconda3/lib/python3.7/site-packages/xgboost_ray/main.py", line 461, in train
    if self._stop_event.is_set():
xgboost_ray.main.RayXGBoostTrainingStopped: Training was interrupted.
2020-12-30 06:35:22,227 INFO main.py:721 -- [RayXGBoost] Created 0 new actors (4 total actors). Waiting until actors are ready for training.
2020-12-30 06:35:22,338 INFO main.py:758 -- [RayXGBoost] Starting XGBoost training.
(pid=9278, ip=172.31.31.127) [06:35:22] task [xgboost.ray]:140179232235216 got new rank 3
(pid=9244, ip=172.31.22.246) [06:35:22] task [xgboost.ray]:139690855271952 got new rank 2
(pid=9436, ip=172.31.17.60) [06:35:22] task [xgboost.ray]:140427761209424 got new rank 0
(pid=9129, ip=172.31.18.124) [06:35:22] task [xgboost.ray]:140607102729744 got new rank 1
2020-12-30 06:35:52,846 INFO main.py:825 -- Training in progress (30 seconds since last restart).
2020-12-30 06:36:22,847 INFO main.py:825 -- Training in progress (60 seconds since last restart).
2020-12-30 06:36:52,847 INFO main.py:825 -- Training in progress (90 seconds since last restart).
2020-12-30 06:37:22,848 INFO main.py:825 -- Training in progress (120 seconds since last restart).
[...]
```

I have just a few questions I'd like to address before we can merge.

File: xgboost_ray/main.py
Comment (krfricke) [inline xgboost_ray/main.py:1003]: This could be problematic. If the actors run into some failure we cannot recover from, this will get into an infinite loop (I encountered a bug during testing in which I ran into this). 
`max_actor_restarts` is linked to non-elastic training, but I'd propose to either a) re-introduce this as a measure to control this behavior or b) introduce another parameter that limits the number of restarts in elastic training. Generally I'd prefer a) to keep the number of parameters low, but would like to hear your opinion on this.

File: xgboost_ray/util.py
Comment (krfricke) [inline xgboost_ray/util.py:184]: This uses internal methods of Ray's global state as there is no public API to get available resources by node. Can we keep this or should I upstream a public API to ray core?

Comment (krfricke) [reply xgboost_ray/util.py:184]: Global cluster resources are not enough to check because we can't know how they are distributed across machines and thus don't know if scheduling an actor is actually feasible.

Comment (richardliaw) [reply xgboost_ray/util.py:184]: instead of using `_check_connected` maybe instead use ray.worker.global_worker.check_connected()?

Comment (richardliaw) [reply xgboost_ray/util.py:184]: _available_resources_per_node is used by serve too -- we should make it public! (ok to merge as is)

File: xgboost_ray/main.py
Comment (richardliaw) [reply xgboost_ray/main.py:1003]: I think max_actor_restarts can be kept for this too.

File: xgboost_ray/util.py
Comment (krfricke) [reply xgboost_ray/util.py:184]: Turns out we need `_check_connected` as it lazily initializes the global state accessor on latest master

**Comment by krfricke** [comment]
I just finished another cluster test with this version, and it seems to be running fine.

**Comment by krfricke** [comment]
Added a PR for the available resource per node here: https://github.com/ray-project/ray/pull/13134

**Comment by krfricke** [comment]
Removed the available resource check. Tested again on cluster, all working. The TestColocation test suite is flaky, I'll raise a separate issue for that.

Should be ready to merge after a final look.

Note that I merged an unrelated benchmarking branch in this PR to allow for more flexible testing.

**Review by richardliaw — APPROVED** [review]

**Pull request merged**
